[flang] Add support for -mrecip[=<list>]

mcinally opened the pull request:
This patch adds support for the -mrecip command line option. The parsing of this options is equivalent to Clang's and it is implemented by setting the "reciprocal-estimates" function attribute.

Also move the ParseMRecip(...) function to CommonArgs, so that Flang is able to make use of it as well.

Commit:
[flang] Add support for -mrecip[=<list>]

This patch adds support for the -mrecip command line option. The parsing of
this options is equivalent to Clang's and it is implemented by setting the
"reciprocal-estimates" function attribute.

Also move the ParseMRecip(...) function to CommonArgs, so that Flang is able to
make use of it as well.

vzakhari reviewed (approved):
Looks great! Thank you, Cameron!

tarunprabhu reviewed (commented):
Thanks for this. Just some minor comments.

mcinally commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -1549,6 +1549,9 @@ LogicalResult ModuleTranslation::convertOneFunction(LLVMFuncOp func) {
   if (auto tuneCpu = func.getTuneCpu())
     llvmFunc->addFnAttr("tune-cpu", *tuneCpu);
 
+  if (auto reciprocalEstimates = func.getReciprocalEstimates())
```
It is `std::optional`:

```
::std::optional< ::llvm::StringRef > getReciprocalEstimates();
```

The others that return an optional `StringRef` also use `auto` as well:

```
::std::optional< ::llvm::StringRef > getTargetCpu();
```

I'll leave this if no objections.

tarunprabhu replied:
I was just wondering because of the `*` operator in the body of the conditional. I think everything else here uses `auto`, so I'd leave it as is.

Commit:
Address a handful of review comments:

1) Rename utility functions. Use camelCase, not PascalCase.
2) Change header comments to be documentation comments.
3) Update Frontend tests to account for implicit attributes.
4) Add Compilation Driver support and test case.

tarunprabhu reviewed (approved):
Thanks for the changes, Cameron. LGTM.

mcinally commented:
> Thanks for the changes, Cameron. LGTM.

Thanks! Would you mind merging yet again? I still don't have access yet.
